lib/helpers/config_builder.py

```
@@ -0,0 +1,110 @@
+import os
+from lib.helpers.yaml_loader import *
+from tools.logger import Logger
+
+log = Logger(os.path.basename(__file__)).new()
+
+
+class ConfigBuilder(YamlLoader):
+
+    # create a singleton pattern
+    def __new__(cls, exec_file):
+        if not hasattr(cls, 'instance'):
+            cls.instance = super(ConfigBuilder, cls).__new__(cls)
+        return cls.instance
+
+    # initialization
+    def __init__(self, dirty_user_data):
+        self.__path_to_envs = 'environments'
+        self.__yaml_helper = YamlLoader
+        self.__dirty_user_data = dirty_user_data
+
+    _config_builder = {}
+
+    # get set configs from ..//environments
+    def _load_all_yaml_configs(self):
+        path_to_dir = os.path.abspath(self.__path_to_envs)
+        return self.__yaml_helper.get_set_dictionaries(self, path_to_dir)
+
+    # merge USER DATA into the CONFIGS
+    def _merge_user_data_to_configs(self):
+        configs_data = self._load_all_yaml_configs()
+        for keys, value in self.__dirty_user_data.items():
+            keys = self._convert_string_keys_to_indexes(keys.split('.'))  # keys - array of keys
+            if len(keys) > 1:
+                self._insert_value_into_config(configs_data, keys, value)  # update configs_data
+        log.info('Merge user data with default yaml configs is successful')
+        return configs_data
+
+    # mix string from USER DATA with configs
+    # insert our value into result dictionary
+    @staticmethod
+    def _insert_value_into_config(configs_data, keys, value):
+        for key in keys[:-1]:                         # take key up to last, equal to [0:last - 1]
+            try:
+                configs_data = configs_data[key]      # nest dict with key into dict
+            except Exception as exc:
```

Do we really need to catch all exceptions?

---

Yes, we need them for catching exceptions when we have some errors in config params.

---

For example: The config: {str(configs_data)} + has not any such keys: "{str(key)}"